Firestore/core/src/model/object_value.cc

```diff
@@ -0,0 +1,287 @@
+      // Insert the accumulated changes at this parent location
+      _google_firestore_v1_MapValue* parent_map = ParentMap(parent);
+      ApplyChanges(parent_map, upserts, deletes);
+      upserts.clear();
+      deletes.clear();
+      parent = path.PopLast();
+    }
+
+    absl::optional<google_firestore_v1_Value> value = data.Get(path);
+    if (value) {
+      upserts.emplace(path.last_segment(), *value);
+    } else {
+      deletes.insert(path.last_segment());
+    }
+  }
+
+  _google_firestore_v1_MapValue* parent_map = ParentMap(parent);
+  ApplyChanges(parent_map, upserts, deletes);
+}
+
+google_firestore_v1_MapValue* MutableObjectValue::ParentMap(
+    const FieldPath& path) {
+  google_firestore_v1_Value* parent = value_.get();
+
+  // Find a or create a parent map entry for `path`.
+  for (const std::string& segment : path) {
+    _google_firestore_v1_MapValue_FieldsEntry* entry =
+        FindEntry(*parent, segment);
+
+    if (entry) {
+      if (entry->value.which_value_type !=
+          google_firestore_v1_Value_map_value_tag) {
+        // Since the element is not a map value, free all existing data and
+        // change it to a map type
+        nanopb::FreeNanopbMessage(google_firestore_v1_Value_fields,
+                                  &entry->value);
+        entry->value.which_value_type = google_firestore_v1_Value_map_value_tag;
+      }
+      parent = &entry->value;
+    } else {
+      // Create a new map value for the current segment
+      _google_firestore_v1_Value new_entry{};
+      new_entry.which_value_type = google_firestore_v1_Value_map_value_tag;
+
+      std::map<std::string, google_firestore_v1_Value> upserts{
+          {segment, new_entry}};
+      ApplyChanges(&(parent->map_value), upserts, {});
+
+      parent = &(FindEntry(*parent, segment)->value);
+    }
+  }
+
+  return &parent->map_value;
+}
+
+void MutableObjectValue::ApplyChanges(
+    google_firestore_v1_MapValue* parent,
+    std::map<std::string, google_firestore_v1_Value> upserts,
+    std::set<std::string> deletes) const {
+  auto source_count = parent->fields_count;
+  auto source_fields = parent->fields;
+
+  // Compute the size of the map after applying all mutations. The final size is
+  // the number of existing entries, plus the number of new entries
+  // minus the number of deleted entries.
+  size_t target_count =
+      upserts.size() +
+      std::count_if(source_fields, source_fields + source_count,
+                    [&](_google_firestore_v1_MapValue_FieldsEntry entry) {
+                      std::string field = nanopb::MakeString(entry.key);
+                      // Check if the entry is deleted or if it is a replacement
+                      // rather than an insert.
+                      return deletes.find(field) == deletes.end() &&
+                             upserts.find(field) == upserts.end();
+                    });
+
+  auto target_fields = static_cast<_google_firestore_v1_MapValue_FieldsEntry*>(
```

Comment:
Can you use `nanopb::MakeArray`? Also, this should be `auto*` to make it clear the variable is a pointer.